cmd/virt-tail/main.go

```diff
@@ -0,0 +1,233 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package main
+
+import (
+	"context"
+	"errors"
+	goflag "flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/nxadm/tail"
+	"github.com/spf13/pflag"
+	"golang.org/x/sync/errgroup"
+
+	"kubevirt.io/client-go/log"
+)
+
+type TermFileError struct{}
+
+func (m *TermFileError) Error() string {
+	return "termFile got detected"
+}
+
+type VirtTail struct {
+	ctx     context.Context
+	logFile string
+	g       *errgroup.Group
+}
+
+func (v *VirtTail) checkFile(socketFile string) bool {
+	_, err := os.Stat(socketFile)
+	return !os.IsNotExist(err)
+}
+
+func (v *VirtTail) tailLogs() error {
+	t, err := tail.TailFile(v.logFile, tail.Config{
+		Follow:        true,
+		CompleteLines: true,
+		MustExist:     false,
+		ReOpen:        true,
+		Logger:        tail.DiscardingLogger,
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		serr := t.Stop()
+		if serr != nil {
+			log.Log.V(3).Infof("tail error: %v", serr)
+		}
+		t.Cleanup()
+	}()
+
+	for {
+		select {
+		case line, ok := <-t.Lines:
+			if !ok {
+				log.Log.V(4).Info("tail error: line not ok")
+			} else if line != nil {
+				if line.Err != nil {
+					log.Log.V(3).Infof("tail error: %v", line.Err)
+				} else {
+					fmt.Println(line.Text)
+				}
+			}
+		case <-v.ctx.Done():
+			return v.ctx.Err()
+		}
+	}
+}
+
+func (v *VirtTail) watchFS() error {
+	socketFile := strings.TrimSuffix(v.logFile, "-log")
+	termFile := v.logFile + "-sigTerm"
+	socketExists := v.checkFile(socketFile)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Log.V(3).Infof("watcher error: %v", err)
+		return err
+	}
+	defer watcher.Close()
+
+	// Add a path.
+	dirPath := filepath.Dir(v.logFile)
+	found := false
+	i := 0
+	for i < 30 && !found {
```

Comment:
I suggest using https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollUntilContextTimeout instead of the for loop, it will simplify a lot.